### ISSUE TITLE
Make defaults explicitly opt-in

### ISSUE THREAD

**Issue by polytypic** [ocaml-multicore/picos#32]
Currently the "defaults" in Picos make the library optionally depend on `unix` and `threads.posix` and `thread-local-storage`.  Unfortunately such optional dependencies do not work quite in the way I'd like them to work.  The defaults also add a significant amount of code to the library.  So, I believe we need to get rid of those dependencies and make it so that the core library builds without them and that one needs to explicitly opt-in to the defaults to use them.  For background, see discussion on defaults [here](https://github.com/ocaml-multicore/picos/discussions/9) and current documentation [here](https://github.com/ocaml-multicore/picos/blob/143de599e3c3a00fb474bb16deacfc04a94256dc/lib/picos/picos.mli#L956-L1043).

**Comment by dinosaure** [comment]
From what I see, system events is mainly required for `cancel_after` which seems a bit specific and not really interesting at first (or, at all, as I think). More importantly, Miou has this specificity that we only need the caml runtime, we don't require any libraries for the core (even `unix.cmxa`). We provide then an API to let the user to implement on its own way and with its libraries the interaction with the system.

This constraint is really important for us if we want to make a scheduler for an unikernel which has probably an other way to notice the user that a system event appeared.

**Comment by polytypic** [comment]
Yes, I definitely plan to drop the dependencies to mentioned libraries from the core.  The dependency to [`backoff`](https://github.com/ocaml-multicore/backoff) should not be too much (it is a tiny library that has no dependencies), but if that is an issue, we could consider inlining it in this case.

On `cancel_after`, I believe timeouts are an important capability for concurrent abstractions.  Pretty much anything that might take a long time should ideally come with support for timeouts (and/or cancelation).  For example, [Kcas has direct support for timeouts](https://github.com/ocaml-multicore/kcas/?tab=readme-ov-file#timeouts) (and it also supports cancelation).

Providing support for (some way to have) timeouts is important to allow libraries (like Kcas) that need them to be [implemented in Picos](https://ocaml-multicore.github.io/picos/doc/picos/Picos/index.html#implemented-in-picos).  In other words, to be able to implement concurrent abstractions that potentially require timeouts on any [Picos compatible](https://ocaml-multicore.github.io/picos/doc/picos/Picos/index.html#picos-compatible) scheduler.

In many older languages you will probably see support for threads and some form of sleep (e.g. OCaml) and perhaps support for canceling (or interrupting) threads, while in some newer runtimes you might see direct support for timeouts independently of threads (e.g. `setTimeout` in browsers and Node.js).  If you have fibers, sleep, and cancelation you can use those to implement timeouts.  Or you can use timeouts to implement [sleep](https://github.com/ocaml-multicore/picos/blob/b881afca3b1e5b844bc0519179dfcf9367c43aec/test/lib/elements/sleep.ml#L1-L23).

In other words, `sleep` and `cancel_after` could be two alternative ways to allow concurrent abstractions that need timeouts to be implemented in Picos.  I chose to provide `cancel_after`, because I believe it is simpler and more efficient primitive as it allows simple timeouts (like in Kcas) to be setup without spawning fibers.
